libcperciva/datastruct/elasticarray.c

```diff
@@ -161,8 +161,7 @@ elasticarray_append(struct elasticarray * EA,
 	size_t nsize;
 
 	/* Check for overflow. */
-	if ((nrec > SIZE_MAX / reclen) ||
-	    (nrec * reclen > SIZE_MAX - EA->size)) {
+	if (nrec > ((SIZE_MAX - EA->size) / reclen)) {
```

Comment:
For the record, these changes to the overflow checks look great.

I was curious, so I dug into the archives.  Those overflow checks were part of the earliest instance of `elasticarray.c` I could find, which was spiped 0.9.9.1 on 2011-07-02.  I'm guessing that Colin copied the overflow check from `elasticarray_resize()` to `_append()` and `_shrink()`, then added a function-specific check to each.

Reply:
I don't know what I was thinking 12 years ago, but that idiom is often used when `reclen` is a constant since it allows the division to be performed at compile-time.  Of course in this case `reclen` isn't a constant, but maybe we could move that part into the macro in `elasticarray.h`?

I don't know if the performance difference would be meaningful though; want to code up a performance test for elasticarray?

Reply:
I think one reason of doing it the division way  i.e., nrec > SIZE_MAX / reclen instead of the multiplication way i.e., nrec * reclen > SIZE_MAX is so that the nrec * reclen doesn't overflow, right?

Reply:
My initial guess is that we're not going to see any different in performance in any real use.  I mean, as soon as `malloc()` or network traffic gets involved, the time it takes arithmetic is going to be a total non-issue.

But sure, I'll write up some perftests.  It would still be interesting to see if it would be faster in a pathological case.

Reply:
@dorjoy03: yes, doing that check with a division is necessary to avoid overflow.

Colin's idea is to move the
```
if ((nrec > SIZE_MAX / reclen)
```
check into the header file, in the definition of `prefix##_append()`.  Since `prefix##_append()` knows the value of `reclen` at compile-time, the compiler can do the division, so that check would be a simple `if (nrec > SOMETHING)`.  If that overflow check doesn't fail, then we'd call `elasticarray_append()` and do the new check.

The question now is "is having a check in the header file worth it?".  I'm leaning towards "no", because I'd expect that overflow check to trigger approximately 0% of the time.  But I need to investigate more, because there might be some cases where it does arise.  And even if it's not worth it for `_append()`, it might be worth it for `_shrink()`.

I'm not working tomorrow, but I'll revisit this on Tuesday.

We're cautious about changing code.  One tech company uses the motto "move fast and break things"; we aim to do the reverse: "move slow and don't break anything".  :)

Reply:
The reason for removing the first condition expression is that if the first condition evaluates to true, the second condition will also evaluate to true. So doing only the second condition check without overflow should handle all the cases that were handled before. 
But of course it's always good to be safer. Let me know if I can be of any help.